[sdk-stamps] add sdk stamps to bundle images

varshaprasad96 opened the pull request:
<!--

Welcome to the Operator SDK! Before contributing, make sure to:

- Read the contributing guidelines https://github.com/operator-framework/operator-sdk/blob/master/CONTRIBUTING.MD
- Rebase your branch on the latest upstream master
- Add a changelog file by copying changelog/fragments/00-template.yaml 
- Link any relevant issues, PR's, or documentation
- When fixing an issue, add "Closes #<ISSUE_NUMBER>"

-->

**Description of the change:**
Add SDK related metric labels and annotations to bundle images

**Motivation for the change:**
To enable OLM track SDK Operators with help of these stamps on bundle resources.

Things to do:

- [X] Add stamps to bundle.Dockerfile and annotaions.yaml
- [x] Add stamps csv (during `generate csv` command)
    - [x] Write test cases to verify the stamps on CSV
- [x] Add stamps to crd (during `generate crds` command)

**Note**:
Adding of labels to CRDs through controller-gen for new project layouts will be done in a new PR. Adding labels through built-in crd-gen for legacy layouts is removed from this PR as of now, and will be included in a follow up later.

Co-authored by: @estroz

varshaprasad96 commented:
@estroz @joelanford @jmrodri Would be helpful to have initial reviews

varshaprasad96 commented on cmd/operator-sdk/bundle/create.go:
```diff
@@ -316,3 +316,65 @@ func copyScorecardConfig() error {
+	return filepath.Join(filepath.Dir(manifestsDir), bundle.MetadataDir, bundle.AnnotationsFile)
+}
+
+func getSDKMetricsContent() map[string]string {
```
Added this function here so that it remains consistent by adding scorecard labels and metrics labels in `create.go`

varshaprasad96 commented on internal/util/projutil/sdk_metrics_util.go:
```diff
@@ -0,0 +1,67 @@
+
+// getSDKProjectLayout returns the `layout` field in PROJECT file if it is a
+// Kubebuilder scaffolded project, or else returns the kind of operator.
+func getSDKProjectLayout() string {
```
TODO: Modify for KB `--project-version` - 1 and 2

estroz commented on internal/util/projutil/project_util.go:
```diff
@@ -367,3 +369,36 @@ func appendContent(fileContents, instruction, content string) (string, error) {
+
+// RewriteAnnotationsYaml unmarshalls the specified yaml file, appends the content and
+// converts it again to yaml.
+func RewriteAnnotationsYaml(filename, directory string, content map[string]string) error {
```
This can go in `internal/util/registry`.

estroz replied:
Correction: `internal/registry`.

estroz commented on internal/util/projutil/project_util.go:
```diff
@@ -367,3 +369,36 @@ func appendContent(fileContents, instruction, content string) (string, error) {
+
+	annotationsYaml := &bundle.AnnotationMetadata{}
+	if err := yaml.Unmarshal(f, annotationsYaml); err != nil {
+		return fmt.Errorf("failed to parse annotations file: %v", err)
```
For consistency
```suggestion
		return fmt.Errorf("error parsing annotations: %v", err)
```

estroz commented on internal/util/projutil/sdk_metrics_util.go:
```diff
@@ -0,0 +1,62 @@
+}
+
+func getSDKBuilder() string {
+	builder := operatorSDK + "-" + ver.GitVersion
```
This can be a `const`. Also
```suggestion
	builder := operatorSDK + "-" + ver.Version
```

estroz commented on internal/util/projutil/sdk_metrics_util.go:
```diff
@@ -0,0 +1,62 @@
+}
+
+func getMediaType() string {
+	return "metrics+v1"
```
This can be a `const`.

estroz commented on internal/util/projutil/sdk_metrics_util.go:
```diff
@@ -0,0 +1,62 @@
+	layout      = "operators.operatorframework.io.metrics.project_layout"
+)
+
+type Sdkmetrics struct {
```
How about
```go
type MetricsLabels struct {
	Data map[string]string
}
```

estroz commented on internal/util/projutil/sdk_metrics_util.go:
```diff
@@ -0,0 +1,62 @@
+	SdkMetricsData map[string]string
+}
+
+func GenerateSDKMetrics() Sdkmetrics {
```
```suggestion
func MakeMetricsLabels() MetricsLabels {
```

estroz commented on internal/util/projutil/sdk_metrics_util.go:
```diff
@@ -0,0 +1,62 @@
+
+const (
+	operatorSDK = "operator-sdk"
+	mediatype   = "operators.operatorframework.io.metrics.mediatype.v1"
```
I have a feeling we'll need these to be public.

varshaprasad96 replied:
Yes, they would. Starting with being private, and will make it public when required.

estroz commented on internal/util/projutil/sdk_metrics_util.go:
```diff
@@ -0,0 +1,62 @@
+// Kubebuilder scaffolded project, or else returns the kind of operator.
+func getSDKProjectLayout() string {
+	if kbutil.HasProjectFile() {
+		cfg, _ := kbutil.ReadConfig()
```
Check this error.

varshaprasad96 commented on internal/generate/olm-catalog/csv_go_test.go:
```diff
@@ -119,13 +123,21 @@ func TestGoCSVNewWithInputsToOutput(t *testing.T) {
+	if err != nil {
+		t.Errorf("error occured %v", err)
+	}
+	assert.Equal(t, expStruct, csvOutputStruct)
```
Manipulating string and removing contents in YAML was causing error. This verbose code could be removed once we have these tests in Ginkgo/gomega format.

varshaprasad96 commented on internal/generate/clusterserviceversion/bases/clusterserviceversion.go:
```diff
@@ -70,6 +70,9 @@ func (b ClusterServiceVersion) GetBase() (base *v1alpha1.ClusterServiceVersion,
 
+	// Add sdk stamps to csv
+	addSDKstamps(base)
+
```
Adding SDK stamps while creating base CSV as these are constant values which user should not change and should be present always.

estroz commented on internal/util/projutil/sdk_metrics_util.go:
```diff
@@ -0,0 +1,78 @@
+)
+
+var (
+	sdkBuilder = OperatorSDK + "-" + ver.GitVersion
```
This version will change depending on what variables were set during `go build` or `go install` of the SDK, ex. `v0.18.0-13-g8a008bc0-dirty`. We should either trim the build info from version and just set `vX.Y.Z+git`.

estroz commented on internal/registry/validate.go:
```diff
@@ -16,17 +16,24 @@ package registry
+	"github.com/operator-framework/operator-registry/pkg/lib/bundle"
 	registrybundle "github.com/operator-framework/operator-registry/pkg/lib/bundle"
 	log "github.com/sirupsen/logrus"
+	"gopkg.in/yaml.v2"
```
```suggestion
	"sigs.k8s.io/yaml"
```

varshaprasad96 replied:
Switching over to `gopkg.in/yaml.v2`. k8s-yaml interprets and appends the field of yaml struct while marshaling and unmarshaling, whereas gopkg-yaml uses the yaml annotation defined in the struct. Example - https://play.golang.org/p/uQ_J4oZBbCR

camilamacedo86 replied:
I just replaced by the `"sigs.k8s.io/yaml"` and the tests worked successfully. Both libs help us work with yaml. However, we are trying to keep all project using `"sigs.k8s.io/yaml"` and remove any other module/dep which has the same goal. Ideally, how much fewer deps we have added to the project would be better to keep it maintained. Did you find any problem with ``"sigs.k8s.io/yaml"`` which justify you use the other lib?

varshaprasad96 replied:
The unit tests do pass with `"sigs.k8s.io/yaml"`. However, it would fail in the scenario where we create bundle image using files already present on disk.
Explanation is as below:

To reproduce the scenario:
1. First create bundle files on disk with `operator-sdk bundle create` command and `--generate-only` flag.
2. Now create bundle image and push it to the registry (for eg. use the command: `operator-sdk bundle create  quay.io/example/memcached-operator:latest --directory path/to/manifests --package memcached-operator`)

Cause of error:
When we use `sigs.k8s.io/yaml`, the key in `annotations.yaml` file would be the name of the field in [`AnnotationMetadata`](https://github.com/operator-framework/operator-registry/blob/master/pkg/lib/bundle/generate.go#L35) struct. The `annotations.yaml` which would be rewritten after adding the sdk stamps, would be like this:
```
Annotations:
  operators.operatorframework.io.bundle.channel.default.v1: stable
  operators.operatorframework.io.bundle.channels.v1: stable
  operators.operatorframework.io.bundle.manifests.v1: manifests/
```

Now when during unmarshalling, ie [here](https://github.com/operator-framework/operator-registry/blob/master/pkg/lib/bundle/generate.go#L265), it would not be able to interpret this because the yaml has the key `Annotations ` not `annotations`. 

If we use `"gopkg.in/yaml.v2"`, the `annotations.yaml` which we re-write would take the marker value defined in yaml struct, `yaml:"annotations"` as it does in operator-registry and it would unmarshall properly. 

I have created a simple example to show the difference between both the packages in this playground: https://play.golang.org/p/uQ_J4oZBbCR

camilamacedo86 replied:
See that your example worked if we replace the yaml for json

> Something int `yaml:"somethingGreat,omitempty"`
With
> Something int `json:"somethingGreat,omitempty"`

Then, see that also it will work with `yaml ` if we replace the `Unmarshal/Marchal` for `JSONToYAML/YAMLToJSON`. See: https://play.golang.org/p/D2guhgb7cvc

In SDK, we read/update/delete the files and use the lib to verify them. In this way, no matter the dep/lib that we use to manipulate it, all should work fine if at the end we have the same result.  Then, I understand that if we change the impl [here](https://github.com/operator-framework/operator-sdk/blob/c4d9a0ef121cee83e0549fce715ee200439b4aad/internal/registry/validate.go#L179-L203) to use the above it should work.  

However, if you still just facing problems with. could you please add a comment with a //todo: replace it for "sigs.k8s.io/yaml". Operator-registry defines the annotations with yaml format.

varshaprasad96 replied:
Adding //todo: because of the dependency on operator-registry. OR defines annotations.

joelanford replied:
This could likely be fixed by adding `json:"annotations"` to the struct tag in the operator-registry repo.

So it would look like: 

```go
type AnnotationMetadata struct {
	Annotations map[string]string `yaml:"annotations" json:"annotations"``
}

```

varshaprasad96 replied:
Will continue with above and have an upstream PR in parallel. Ill have a follow-up PR to remove the TODO.

estroz commented on internal/generate/olm-catalog/csv_go_test.go:
```diff
@@ -472,3 +528,52 @@ func TestUpdateCSVVersion(t *testing.T) {
+
+// removeSDKstampsFromCSV removes the sdk stamps from CSV struct. Used for test cases where
+// we need to test the generated CSV with expected predefined CSV file on disk.
+func removeSDKstampsFromCSVHelper(csv *v1alpha1.ClusterServiceVersion) {
```
This could be done more easily for tests by regexp replacing annotations on CSV bytes. Then you don't have to re-marshal CSV's in each test.

varshaprasad96 replied:
I had tried using regex to do this. But manipulating yaml in string format and making a string comparison again seems error prone due to spaces.

estroz replied:
On second thought this is fine for now. Legacy code will be removed soon anyway.

varshaprasad96 commented on cmd/operator-sdk/bundle/create.go:
```diff
@@ -230,11 +232,21 @@ func (c bundleCreateCmd) runGenerate() error {
 		c.overwrite = true
 	}
 
+	annotationsFile := getAnnotationsFilePath(c.outputDir, c.directory)
```
If there is a pre-existing `annotations.yaml`, the `bundle.ValidateAnnotation` inside `GenerateFunc` unmarshals the file and checks the length of entries in the map which is supposed to be 6. In order to pass the validation, we would have to remove any external entries.

estroz replied:
These labels need to be added to the `bundle.Dockerfile` as well. There shouldn't be a mismatch between annotations and labels, even though we aren't using docker labels, right @kevinrizza @dinhxuanvu?

dinhxuanvu replied:
Correct. They should match.

camilamacedo86 commented on internal/generate/olm-catalog/csv_go_test.go:
```diff
@@ -15,21 +15,25 @@
 	"path/filepath"
 	"testing"
 
-	"github.com/operator-framework/operator-sdk/internal/generate/clusterserviceversion/bases"
+	"github.com/blang/semver"
+	"github.com/operator-framework/api/pkg/operators/v1alpha1"
+	olmapiv1alpha1 "github.com/operator-framework/api/pkg/operators/v1alpha1"
 	"github.com/operator-framework/operator-sdk/internal/util/fileutil"
 	"github.com/operator-framework/operator-sdk/internal/util/k8sutil"
 	"github.com/operator-framework/operator-sdk/internal/util/projutil"
+	log "github.com/sirupsen/logrus"
+	"sigs.k8s.io/yaml"
 
-	"github.com/blang/semver"
-	"github.com/operator-framework/api/pkg/operators/v1alpha1"
+	"github.com/operator-framework/operator-sdk/internal/generate/clusterserviceversion/bases"
 	"github.com/stretchr/testify/assert"
-	"sigs.k8s.io/yaml"
```
Missing sorts the imports properly. 

```go
	"bytes"
	"fmt"
	"io/ioutil"
	"os"
	"os/exec"
	"path/filepath"
	"testing"

	"github.com/blang/semver"
	"github.com/operator-framework/api/pkg/operators/v1alpha1"
	olmapiv1alpha1 "github.com/operator-framework/api/pkg/operators/v1alpha1"
	log "github.com/sirupsen/logrus"
	"sigs.k8s.io/yaml"
	"github.com/stretchr/testify/assert"


	"github.com/operator-framework/operator-sdk/internal/generate/clusterserviceversion/bases"
        "github.com/operator-framework/operator-sdk/internal/util/fileutil"
	"github.com/operator-framework/operator-sdk/internal/util/k8sutil"
	"github.com/operator-framework/operator-sdk/internal/util/projutil"
)

```

camilamacedo86 commented on internal/util/projutil/sdk_stamps_util_test.go:
```diff
@@ -0,0 +1,39 @@
+)
+
+var _ = Describe("Testing sdk stamps helper function", func() {
+	Describe("Test sdk_metrics_util helpers", func() {
```
```suggestion
```

camilamacedo86 commented on internal/util/projutil/sdk_stamps_util_test.go:
```diff
@@ -0,0 +1,39 @@
+			})
+
+		})
+	})
```
```suggestion
```

camilamacedo86 commented on internal/util/projutil/sdk_stamps_util.go:
```diff
@@ -0,0 +1,85 @@
+)
+
+const (
+	OperatorSDK      = "operator-sdk"
+	BundleMediatype  = "operators.operatorframework.io.metrics.mediatype.v1"
+	BundleSDKBuilder = "operators.operatorframework.io.metrics.builder"
+	BundleSDKLayout  = "operators.operatorframework.io.metrics.project_layout"
+	sdkMediatype     = "metrics+v1"
+	OperatorBuilder  = "operators.operatorframework.io/builder"
+	OperatorLayout   = "operators.operatorframework.io/project_layout"
```
We just put const in upper case when we need to export the value. 
Also, we create global vars when the values are used in more than one place. 

```suggestion
	OperatorBuilder  = "operators.operatorframework.io/builder"
	OperatorLayout   = "operators.operatorframework.io/project_layout"
```

camilamacedo86 commented on internal/util/projutil/sdk_stamps_util.go:
```diff
@@ -0,0 +1,85 @@
+	OperatorLayout   = "operators.operatorframework.io/project_layout"
+)
+
+type MetricLabels struct {
+	Data map[string]string
+}
+
+// MakeBundleMetricsLabels returns the SDK metric stamps which will be added
+// to bundle resources like bundle.Dockerfile and annotations.yaml.
+func MakeBundleMetricsLabels() MetricLabels {
+	return MetricLabels{
+		Data: map[string]string{
+			BundleMediatype:  sdkMediatype,
+			BundleSDKBuilder: getSDKBuilder(),
+			BundleSDKLayout:  getSDKProjectLayout(),
+		},
+	}
+}
```
has no need to create a structure object when has nothing to represent. See that you are adding just Data and a map, and then adding different kinds of information on it. So, we can just: 

```suggestion
// MakeBundleMetricsLabels returns the SDK metric stamps which will be added
// to bundle resources like bundle.Dockerfile and annotations.yaml.
func MakeBundleMetricsLabels() map[string]string {
	return map[string]string{
		"operators.operatorframework.io.metrics.mediatype.v1":  "metrics+v1",
		"operators.operatorframework.io.metrics.builder": getSDKBuilder(),
		"operators.operatorframework.io.metrics.project_layout":  getSDKProjectLayout(),
	}
}
```

Also, see that it will simplifie the place where it is called/used.

camilamacedo86 commented on internal/util/projutil/sdk_stamps_util.go:
```diff
@@ -0,0 +1,85 @@
+
+// MakeOperatorMetricLables returns the SDK metric stamps which will be added
+// to custom resource definitions and cluster service versions.
+func MakeOperatorMetricLables() MetricLabels {
+	return MetricLabels{
+		Data: map[string]string{
+			OperatorBuilder: getSDKBuilder(),
+			OperatorLayout:  getSDKProjectLayout(),
+		},
+	}
+}
```
```suggestion
// MakeOperatorMetricLables returns the SDK metric stamps which will be added
// to custom resource definitions and cluster service versions.
func MakeOperatorMetricLables() map[string]string {
	return map[string]string{
			OperatorBuilder: getSDKBuilder(),
			OperatorLayout:  getSDKProjectLayout(),
	}
}
```

camilamacedo86 commented on internal/util/projutil/sdk_stamps_util.go:
```diff
@@ -0,0 +1,85 @@
+}
+
+func getSDKBuilder() string {
+	return OperatorSDK + "-" + parseVersion(ver.GitVersion)
```
```suggestion
	return "operator-sdk" + "-" + parseVersion(ver.GitVersion)
```
The value is not used in another place because we do not need a global var.

camilamacedo86 commented on internal/registry/validate.go:
```diff
@@ -127,3 +131,75 @@ func appendResult(results []apierrors.ManifestResult, r apierrors.ManifestResult
+	if err != nil {
+		return err
+	}
+
+	metricsContent := projutil.MakeBundleMetricsLabels()
+
+	// Delete the contents to annotationsYaml
+	for key := range metricsContent.Data {
+		delete(metadata.Annotations, key)
+	}
+
```
```suggestion
	// Delete the contents to annotationsYaml
	for key := range  projutil.MakeBundleMetricsLabels() {
		delete(metadata.Annotations, key)
	}
```

camilamacedo86 commented on internal/generate/clusterserviceversion/bases/clusterserviceversion.go:
```diff
@@ -89,6 +92,17 @@ func (b ClusterServiceVersion) GetBase() (base *v1alpha1.ClusterServiceVersion,
+// stamp is not present, then sdk stamps are also not populated in csv.
+func addSDKstamps(csv *v1alpha1.ClusterServiceVersion) {
+	if _, exist := csv.ObjectMeta.Annotations[projutil.OperatorBuilder]; !exist {
+		metricLabels := projutil.MakeOperatorMetricLables()
+		for label, value := range metricLabels.Data {
+			csv.ObjectMeta.Annotations[label] = value
+		}
```
```suggestion
		for label, value := range  projutil.MakeOperatorMetricLables() {
			csv.ObjectMeta.Annotations[label] = value
		}
```

camilamacedo86 commented on internal/generate/crd/crd.go:
```diff
@@ -371,6 +374,27 @@ func setCRDStorageVersion(crd *apiextv1beta1.CustomResourceDefinition) {
+}
+
+func addStampsToAnnotations(mapAnnotations map[string]string, operatorType string) {
+	metricLables := projutil.MakeOperatorMetricLables()
+	for label, value := range metricLables.Data {
+		mapAnnotations[label] = value
+	}
```
```suggestion
	for label, value := range projutil.MakeOperatorMetricLables() {
		mapAnnotations[label] = value
	}
```

camilamacedo86 commented on internal/generate/olm-catalog/csv_go_test.go:
```diff
@@ -472,3 +528,52 @@ func TestUpdateCSVVersion(t *testing.T) {
+// we need to test the generated CSV with expected predefined CSV file on disk.
+func removeSDKstampsFromCSVHelper(csv *v1alpha1.ClusterServiceVersion) {
+	if _, exist := csv.ObjectMeta.Annotations[projutil.OperatorBuilder]; exist {
+		metricLabels := projutil.MakeOperatorMetricLables()
+		for label := range metricLabels.Data {
+			delete(csv.ObjectMeta.Annotations, label)
+		}
```
```suggestion
		for label := range projutil.MakeOperatorMetricLables() {
			delete(csv.ObjectMeta.Annotations, label)
		}
```

camilamacedo86 commented on cmd/operator-sdk/bundle/create.go:
```diff
@@ -330,3 +342,68 @@ func copyScorecardConfig() error {
+	annotationsFilePath := getAnnotationsFilePath(outputDir, manifestsDir)
+
+	// Write metrics annotations
+	metricsContent := projutil.MakeBundleMetricsLabels()
+	err := registry.RewriteAnnotationsYaml(annotationsFilePath, metricsContent.Data)
+	if err != nil {
+		return err
+	}
```
```suggestion
	err := registry.RewriteAnnotationsYaml(annotationsFilePath, projutil.MakeBundleMetricsLabels())
	if err != nil {
		return err
	}
```

camilamacedo86 commented on cmd/operator-sdk/bundle/create.go:
```diff
@@ -330,3 +342,68 @@ func copyScorecardConfig() error {
+	return filepath.Join(filepath.Dir(manifestsDir), bundle.MetadataDir, bundle.AnnotationsFile)
+}
+
+func getSDKMetricsContent() map[string]string {
+	content := projutil.MakeBundleMetricsLabels()
+	return content.Data
+}
```
The getSDKMetricsContent() is not required because we can use directly projutil.MakeBundleMetricsLabels()
```suggestion
```

camilamacedo86 commented on cmd/operator-sdk/bundle/create.go:
```diff
@@ -330,3 +342,68 @@ func copyScorecardConfig() error {
+func addSDKMetricLabels() error {
+	var sdkMetricContent string
+
+	content := getSDKMetricsContent()
+
+	for key, value := range content {
+		sdkMetricContent += fmt.Sprintf("LABEL %s=%s\n", key, value)
+	}
```
```suggestion
	for key, value := range projutil.MakeBundleMetricsLabels() {
		sdkMetricContent += fmt.Sprintf("LABEL %s=%s\n", key, value)
	}
```

camilamacedo86 commented on cmd/operator-sdk/bundle/create.go:
```diff
@@ -330,3 +342,68 @@ func copyScorecardConfig() error {
+
+func addSDKMetricLabels() error {
+	var sdkMetricContent string
+
```
Could we remove the spaces between the code lines? 
They are not helpful to understand the code.

camilamacedo86 commented on internal/generate/crd/crd_test.go:
```diff
@@ -283,6 +290,9 @@ spec:
+  annotations:
+    operators.operatorframework.io/builder: operator-sdk-unknown
+    operators.operatorframework.io/project_layout: unknown
   name: memcacheds.cache.example.com
```
Unfortunately, I have no opportunity to check the doc to raise it up and I still think that might would be better we follow up the current layout. However, since it was approved in the enhancement doc let's move forward with. You are right 👍

varshaprasad96 commented:
Thanks @camilamacedo86 ! It was really helpful. I have made the suggested changes and also addressed the comment regarding `gopkg.in/yaml.v2`. Please have a look at it.

camilamacedo86 commented on internal/util/projutil/sdk_stamps_util.go:
```diff
@@ -0,0 +1,76 @@
+	OperatorLayout  = "operators.operatorframework.io/project_layout"
+)
+
+type MetricLabels struct {
+	Data map[string]string
+}
```
Has no need for we have the structure here. 
```suggestion
```

camilamacedo86 commented on internal/generate/crd/crd_test.go:
```diff
@@ -231,6 +232,9 @@ func TestCRDNonGo(t *testing.T) {
 const crdNonGoDefaultExpV1beta1 = `apiVersion: apiextensions.k8s.io/v1beta1
 kind: CustomResourceDefinition
 metadata:
+  annotations:
```
Yep, it is fine. You are right.

camilamacedo86 commented on cmd/operator-sdk/add/api.go:
```diff
@@ -187,7 +187,7 @@ func doAnsibleAPIScaffold() error {
 	if err != nil {
 		return fmt.Errorf("new ansible api scaffold failed: %v", err)
```
Should we not add a fragment for this change?

estroz commented on cmd/operator-sdk/bundle/create.go:
```diff
@@ -230,11 +232,21 @@ func (c bundleCreateCmd) runGenerate() error {
+	annotationsFile := getAnnotationsFilePath(c.outputDir, c.directory)
+
+	if isExist(annotationsFile) && !c.overwrite {
+		err := registry.RemoveSDKstampsInAnnotation(annotationsFile)
```
You shouldn't need to remove these if you're writing labels to `bundle.Dockerfile` too.

estroz replied:
Fixed by https://github.com/operator-framework/operator-registry/pull/358.

camilamacedo86 commented on internal/generate/crd/crd.go:
```diff
@@ -273,6 +275,7 @@ func (g Generator) generateNonGo() (map[string][]byte, error) {
 
 	sort.Sort(k8sutil.CRDVersions(crd.Spec.Versions))
 	setCRDStorageVersion(crd)
+	setSDKstamps(crd, g.OperatorType)
```
```suggestion
	g.setSDKstamps(crd)
```

estroz replied:
This _could_ be a method, but is more easily unit-tested if it isn't.

camilamacedo86 commented on internal/generate/crd/crd.go:
```diff
@@ -371,6 +374,26 @@ func setCRDStorageVersion(crd *apiextv1beta1.CustomResourceDefinition) {
 	crd.Spec.Versions[0].Storage = true
 }
 
+func setSDKstamps(crd *apiextv1beta1.CustomResourceDefinition, operatorType string) {
```
```suggestion
func (g Generator) setSDKstamps(crd *apiextv1beta1.CustomResourceDefinition) {
```
I think is better than pass the type.

camilamacedo86 commented on internal/generate/crd/crd.go:
```diff
@@ -371,6 +374,26 @@ func setCRDStorageVersion(crd *apiextv1beta1.CustomResourceDefinition) {
+	crd.SetAnnotations(annotations)
+}
+
+func addStampsToAnnotations(mapAnnotations map[string]string, operatorType string) {
```
```suggestion
func (g Generator) addStampsToAnnotations(mapAnnotations map[string]string) {
```
same here ^

estroz replied:
This _could_ be a method, but is more easily unit-tested if it isn't.

camilamacedo86 commented on internal/generate/crd/crd.go:
```diff
@@ -185,9 +188,8 @@ func (g Generator) generateGo() (map[string][]byte, error) {
-				annotations = nil
-			}
+			// Add sdk related stamps to the cached annotations
+			addStampsToAnnotations(annotations, projutil.OperatorTypeGo)
```
```suggestion
			g.addStampsToAnnotations(annotations)
```

camilamacedo86 commented on internal/generate/crd/crd.go:
```diff
@@ -371,6 +374,26 @@ func setCRDStorageVersion(crd *apiextv1beta1.CustomResourceDefinition) {
+	}
+	// Modifying operator type for ansible and helm legacy operators, as during
+	// scaffolding the project directories are not written on disk
+	if operatorType != projutil.OperatorTypeGo {
```
```suggestion
	if g.operatorType != projutil.OperatorTypeGo {
```

camilamacedo86 commented on internal/generate/crd/crd.go:
```diff
@@ -371,6 +374,26 @@ func setCRDStorageVersion(crd *apiextv1beta1.CustomResourceDefinition) {
+	if annotations == nil {
+		annotations = make(map[string]string)
+	}
+	addStampsToAnnotations(annotations, operatorType)
```
```suggestion
	g.addStampsToAnnotations(annotations)
```

camilamacedo86 commented on internal/generate/crd/crd.go:
```diff
@@ -371,6 +374,26 @@ func setCRDStorageVersion(crd *apiextv1beta1.CustomResourceDefinition) {
 }
 
+func setSDKstamps(crd *apiextv1beta1.CustomResourceDefinition, operatorType string) {
+	annotations := crd.GetAnnotations()
+	if annotations == nil {
+		annotations = make(map[string]string)
+	}
```
```suggestion
	if len(crd.GetAnnotations()) < 1 {
		annotations = make(map[string]string)
	}
```

estroz replied:
The only situation in which we need to create a map is if `crd.GetAnnotations() == nil`. Otherwise a non-nil map exists and can be accessed.

varshaprasad96 commented:
@estroz Modified the PR based on bumping of OR version. Please take a look.

estroz commented on cmd/operator-sdk/bundle/create.go:
```diff
@@ -330,3 +333,57 @@ func copyScorecardConfig() error {
+	return nil
+}
+
+func rewriteBundleImageContents(outputDir, manifestDir string) error {
```
A more consistent logical flow:
```go
func rewriteBundleImageContents(rootDir string) error {
    metricsLabels := projutil.MakeBundleMetricsLabels()

	// Write metrics labels to bundle.Dockerfile.
    if err := addLabelsToDockerfile(bundle.Dockerfile, metricsLabels); err != nil {
        return fmt.Errorf("error writing bundle image scorecard labels: %v", err)
    }

	// Write metrics annotations to annotations.yaml.
    annotationsFilePath := getAnnotationsFilePath(rootDir)
	err := addLabelsToAnnotations(annotationsFilePath, metricsLabels)
	if err != nil {
        return fmt.Errorf("error writing bundle scorecard annotations: %v", err)
	}

	// CopyScorecardConfig to bundle.Dockerfile. Do this last to ensure that the above operations were successful.
	if err := copyScorecardConfig(bundle.Dockerfile); err != nil {
		return fmt.Errorf("error adding image COPY directive for scorecard config: %v", err)
	}
return nil
```

I've renamed `writeMetricsToAnnotationsYaml -> addLabelsToAnnotations` and `addSDKMetricLabels -> addLabelsToDockerfile` for consistency.

estroz commented on cmd/operator-sdk/bundle/create.go:
```diff
@@ -234,7 +236,8 @@ func (c bundleCreateCmd) runGenerate() error {
 	}
-	if err = copyScorecardConfig(); err != nil {
+
+	if err = rewriteBundleImageContents(c.outputDir, c.directory); err != nil {
```
```suggestion
	rootDir := c.outputDir
	if rootDir == "" {
		rootDir = filepath.Dir(c.directory)
	}
	if err = rewriteBundleImageContents(rootDir); err != nil {
```

estroz commented on internal/generate/clusterserviceversion/bases/clusterserviceversion.go:
```diff
@@ -89,6 +92,16 @@ func (b ClusterServiceVersion) GetBase() (base *v1alpha1.ClusterServiceVersion,
 	return base, nil
 }
 
+// addSDKstamps adds SDK stamps to the base if they do not exist. It assumes that if sdk Builder
```
```go
// setSDKAnnotations adds SDK annotations to csv if they do not exist.
func addSDKstamps(csv *v1alpha1.ClusterServiceVersion) {
	annotations := csv.GetAnnotations()
	if annotations == nil {
		annotations = make(map[string]string)
	}
	for key, value := range projutil.MakeOperatorMetricLables() {
		if _, exist := annotations[key]; !exist {
			annotations[key] = value
		}
	}
	csv.SetAnnotations(annotations)
}
```

estroz commented on internal/util/projutil/sdk_stamps_util.go:
```diff
@@ -0,0 +1,72 @@
+
+// MakeOperatorMetricLables returns the SDK metric stamps which will be added
+// to custom resource definitions and cluster service versions.
+func MakeOperatorMetricLables() map[string]string {
```
```suggestion
func MakeOperatorMetricLabels() map[string]string {
```

estroz commented on internal/util/projutil/sdk_stamps_util.go:
```diff
@@ -0,0 +1,72 @@
+}
+
+func parseVersion(input string) string {
+	re := regexp.MustCompile("v[0-9]*.[0-9]*.[0-9]*")
```
```suggestion
	re := regexp.MustCompile("v[1-9][0-9]+\\.[0-9]+\\.[0-9]+")
```

estroz commented on internal/util/projutil/sdk_stamps_util.go:
```diff
@@ -0,0 +1,72 @@
+}
+
+func getSDKBuilder() string {
+	return "operator-sdk" + "-" + parseVersion(ver.GitVersion)
```
This won't return the right version strictly speaking, since an `operator-sdk` binary at version `v0.18.0-38-g07d4336d-dirty` is not at version `v0.18.0`. Post-release versions would be better represented as `v0.18.x` or `v0.18.0+git` so metrics can detect that the label isn't from a released version, but without having realized git data in the value.

varshaprasad96 replied:
As the unreleased versions will have `dirty` in their git-version, have used it to represent them with `+git` suffix.

estroz replied:
`dirty` means there are files not committed to a branch, so you'd append `+git` if there is any suffix after the patch version (ex. version matches `v[0-9]+\.[0-9]+\.[0-9]+-.+`).

estroz commented on internal/util/projutil/sdk_stamps_util.go:
```diff
@@ -0,0 +1,72 @@
+	if kbutil.HasProjectFile() {
+		cfg, err := kbutil.ReadConfig()
+		if err != nil {
+			return "Project Layout cannot be found"
```
This should be an error if config cannot be read.

varshaprasad96 replied:
Instead of erroring, wouldn't it be reasonable to stamp is as "unknown" (or any other error value), as we do want the other stamps to appear.

estroz commented on internal/util/projutil/sdk_stamps_util.go:
```diff
@@ -0,0 +1,72 @@
+// to bundle resources like bundle.Dockerfile and annotations.yaml.
+func MakeBundleMetricsLabels() map[string]string {
+	return map[string]string{
+		"operators.operatorframework.io.metrics.mediatype.v1":   "metrics+v1",
```
Make these label keys constants.

varshaprasad96 commented:
@estroz Modified the PR based on comments. We had earlier discussed [here](https://docs.google.com/document/d/1E76ajOb7J2bvJb4_r8LyAspd7_rxBVUGa1M5Sdv37h8/edit#) that we would still like to have CRD-stamps on legacy projects and would be injecting it via the built in crd generation. Is it fine to leave it here and have a separate PR for new project layouts?

estroz commented on cmd/operator-sdk/bundle/create.go:
```diff
@@ -330,3 +339,49 @@ func copyScorecardConfig() error {
 }
+
+func addLabelsToDockerfile(filename string, metricAnnotation map[string]string) error {
+	var sdkMetricContent string
```
Nit: use [`strings.Builder`](https://golang.org/pkg/strings/#Builder).

varshaprasad96 commented:
@estroz Removed labels from CRDs. Please have a look.

estroz commented on cmd/operator-sdk/bundle/create.go:
```diff
@@ -330,3 +339,49 @@ func copyScorecardConfig() error {
+		return fmt.Errorf("error writing metric labels to annotations.yaml: %v", err)
+	}
+
+	// CopyScorecardConfig to bundle.Dockerfile
```
```suggestion
	// Add a COPY for the scorecard config to bundle.Dockerfile.
```

estroz commented on internal/generate/clusterserviceversion/bases/clusterserviceversion.go:
```diff
@@ -89,6 +92,22 @@ func (b ClusterServiceVersion) GetBase() (base *v1alpha1.ClusterServiceVersion,
+		annotations = make(map[string]string)
+	}
+
+	if _, exist := annotations[projutil.OperatorBuilder]; !exist {
+		for key, value := range projutil.MakeOperatorMetricLabels() {
+			annotations[key] = value
+		}
+	}
```
```suggestion
	for key, value := range projutil.MakeOperatorMetricLables() {
		if _, exist := annotations[key]; !exist {
			annotations[key] = value
		}
	}
```

estroz commented on internal/generate/clusterserviceversion/clusterserviceversion_test.go:
```diff
@@ -234,28 +265,48 @@ var _ = Describe("Generating a ClusterServiceVersion", func() {
-				csv, err := g.generate()
-				Expect(err).ToNot(HaveOccurred())
-				Expect(csv).To(Equal(baseCSV))
+		Context("to create a new", func() {
```
This unit test format is a reversion to a prior state, please revert.

estroz commented on internal/generate/clusterserviceversion/bases/clusterserviceversion.go:
```diff
@@ -70,6 +70,9 @@ func (b ClusterServiceVersion) GetBase() (base *v1alpha1.ClusterServiceVersion,
 	}
 
+	// Add sdk labels to csv
+	setSDKAnnotations(base)
```
This needs to be moved out of the `bases` package and into the CSV generator itself so the stamps aren't added to CSVs generated by `generate csv`.

estroz commented on internal/generate/clusterserviceversion/clusterserviceversion_test.go:
```diff
@@ -453,3 +506,19 @@ func upgradeCSV(csv *v1alpha1.ClusterServiceVersion, name, version string) *v1al
+// removeSDKLabelsFromCSVString to remove the sdk labels from test CSV structs. The test
+// cases do not generate a PROJECTFILE or an entire operator to get the version or layout
+// of SDK. Hence the values of those will appear "unknown".
+func removeSDKLabelsFromCSVString(csv string) string {
```
You can call this directly from `readFileHelper` so you don't have to add it to every test case.

estroz commented on internal/generate/clusterserviceversion/clusterserviceversion_test.go:
```diff
@@ -453,3 +506,19 @@ func upgradeCSV(csv *v1alpha1.ClusterServiceVersion, name, version string) *v1al
+}
+
+func removeSDKLabelsFromCSVHelper(csv *v1alpha1.ClusterServiceVersion) {
+	if _, exist := csv.ObjectMeta.Annotations[projutil.OperatorBuilder]; exist {
+		for label := range projutil.MakeOperatorMetricLabels() {
+			delete(csv.ObjectMeta.Annotations, label)
+		}
+	}
```
```suggestion
	for label := range projutil.MakeOperatorMetricLabels() {
		delete(csv.ObjectMeta.Annotations, label)
	}
```

varshaprasad96 commented:
Modified the PR addressing the comments. `--generate-csv` command will not add sdk metric labels.

estroz commented on internal/generate/clusterserviceversion/clusterserviceversion.go:
```diff
@@ -160,13 +161,32 @@ func (g *Generator) Generate(cfg *config.Config, opts ...Option) (err error) {
+	}
+
+	for key, value := range projutil.MakeOperatorMetricLabels() {
+		if _, exist := annotations[key]; !exist {
```
With this existence check, consider the following scenario: I built an operator with v0.18.1 of the SDK, and now I've upgraded to v0.19.0. My CSV still has the annotation `operators.operatorframework.io/builder: operator-sdk-v0.18.1`. The same situation is true for `operators.operatorframework.io/project_layout` when we've bumped plugin versions _or_ when a user migrates their operator from Helm to Ansible or Ansible to Go.

If the generator does not overwrite these annotations, we have broken metrics values. If it does, on-disk files are updated with each SDK version bump (I don't see a problem with this necessarily). I'd prefer to overwrite these values. @joelanford @shawn-hurley FYI.

varshaprasad96 replied:
Agreed. Will revert back to overwriting of values.

estroz commented on internal/generate/clusterserviceversion/clusterserviceversion_test.go:
```diff
@@ -134,8 +141,29 @@ var _ = Describe("Generating a ClusterServiceVersion", func() {
 				Expect(outputFile).To(BeAnExistingFile())
-				Expect(string(readFileHelper(outputFile))).To(MatchYAML(baseCSVUIMetaStr))
+
+				annotations := outputCSV.ObjectMeta.Annotations
```
Nit
```suggestion
				annotations := outputCSV.GetAnnotations()
```

estroz commented on internal/generate/clusterserviceversion/clusterserviceversion_test.go:
```diff
@@ -150,8 +178,9 @@ var _ = Describe("Generating a ClusterServiceVersion", func() {
 				}
 				Expect(g.Generate(cfg, opts...)).ToNot(HaveOccurred())
 				outputFile := filepath.Join(tmp, bundle.ManifestsDir, makeCSVFileName(operatorName))
+				outputFileContents := readFileHelper(outputFile)
```
Nit: any reason to not continue inlining `Expect(readFileHelper(...))`?

estroz commented on internal/registry/validate.go:
```diff
@@ -127,3 +130,54 @@ func appendResult(results []apierrors.ManifestResult, r apierrors.ManifestResult
+		return err
+	}
+
+	err = ioutil.WriteFile(filename, []byte(file), 0644)
```
Nit: use the existing mode if possible and default to rw since this file is typically generated with those perms.
```suggestion
	mode := os.FileMode(0666)
	if info, err := os.Stat(filename); err == nil {
		mode = info.Mode()
	}
	err = ioutil.WriteFile(filename, []byte(file), mode)
```

estroz commented on internal/util/projutil/sdk_stamps_util.go:
```diff
@@ -0,0 +1,85 @@
+// to bundle resources like bundle.Dockerfile and annotations.yaml.
+func MakeBundleMetricsLabels() map[string]string {
+	return map[string]string{
+		bundleMediaType: "metrics+v1",
```
Make `"metrics+v1"` a constant

estroz commented on internal/util/projutil/sdk_stamps_util.go:
```diff
@@ -0,0 +1,85 @@
+	if kbutil.HasProjectFile() {
+		cfg, err := kbutil.ReadConfig()
+		if err != nil {
+			return err.Error()
```
```suggestion
			log.Debugf("Error reading config: %v", err)
			return "unknown"
```

estroz commented on internal/util/projutil/sdk_stamps_util_test.go:
```diff
@@ -0,0 +1,41 @@
+	. "github.com/onsi/gomega"
+)
+
+var _ = Describe("Testing sdk labels helper function", func() {
```
```suggestion
var _ = Describe("SDK Label helper functions", func() {
```

estroz commented on internal/util/projutil/sdk_stamps_util_test.go:
```diff
@@ -0,0 +1,41 @@
+)
+
+var _ = Describe("Testing sdk labels helper function", func() {
+	Describe("Testing SDK version", func() {
```
```suggestion
	Describe("parseVersion", func() {
```

estroz commented on cmd/operator-sdk/bundle/create.go:
```diff
@@ -234,7 +236,14 @@ func (c bundleCreateCmd) runGenerate() error {
 	}
-	if err = copyScorecardConfig(); err != nil {
+
+	// rootDir is the location where metadata directory is present.
+	rootDir := c.outputDir
+	if rootDir == "" {
+		rootDir = filepath.Dir(c.directory)
+	}
+
+	if err = rewriteBundleImageContents(rootDir); err != nil {
```
This needs to go into `generate bundle` code as well.

varshaprasad96 replied:
Added to `generate bundle`.

estroz commented on cmd/operator-sdk/generate/bundle/bundle.go:
```diff
@@ -259,5 +260,15 @@ func (c bundleCmd) generateMetadata(manifestsDir, outputDir string) error {
 	if err != nil {
 		return fmt.Errorf("error generating bundle metadata: %v", err)
 	}
+	if c.overwrite {
```
If these files to not exist, labels/annotations should be written too. You can copy relevant code from `bundle create` here since that command will be deprecated and removed at some point soon.

varshaprasad96 replied:
`generateFunc` would create these files and write them, and if `--overwrite` is true, these annotations would be rewritten. The files are being created before this if they do not exist.

estroz replied:
What happens if `--overwrite=false` but these files did not exist before `GenerateFunc` wrote them? These annotations/labels would not be written.

varshaprasad96 replied:
Will modify this to write annotations when `--overwrite=false` and files weren't present before. And to confirm, we would not like the annotations being rewritten when the metadata is already present and overwrite is explicitly set to be `false`.

estroz replied:
Correct

estroz commented on cmd/operator-sdk/generate/bundle/bundle.go:
```diff
@@ -255,9 +256,39 @@ func (c bundleCmd) runMetadata() error {
+		annotationsDir = outputDir + bundle.MetadataDir
+	}
+
+	if genutil.IsNotExist(bundle.DockerFile) || genutil.IsNotExist(annotationsDir) {
```
To make this check robust, you need to check and write to each file individually, since one may exist without the other.

varshaprasad96 replied:
Yes, but I was thinking of rewriting both of them even if one exists, because until tweaked manually `GenerateFunc` of OR is going to write both and also `generate bundle` provides a `--metadata` option which writes both. In the sense, if there is a scenario where only one of them exists, doesn't it mean that either it has been manually deleted there is some error (though I assume its taken care of in respective commands `create/generate`) and wouldn't rewriting **both** of them together in a clean state be a better option in such a scenario?
